source/slang/slang.cpp

```diff
@@ -1725,14 +1727,51 @@ void Module::setModuleDecl(ModuleDecl* moduleDecl)
     m_moduleDecl = moduleDecl;
 }
 
-// ComponentType
+RefPtr<EntryPoint> Module::findEntryPointByName(UnownedStringSlice const& name)
+{
+    // TODO: We should consider having this function be expanded to be able
+    // to look up and validate possible entry-point functions in teh module
+    // even if they were not marked with `[shader(...)]` in the source code.
+    //
+    // With such a change the function would probably need to accept a stage
+    // to use and a sink to write validation errors to.
 
-static const Guid IID_IComponentType = SLANG_UUID_IComponentType;
+    for(auto entryPoint : m_entryPoints)
+    {
+        if(entryPoint->getName()->text.getUnownedSlice() == name)
+            return entryPoint;
+    }
+
+    return nullptr;
+}
+
+void Module::_addEntryPoint(EntryPoint* entryPoint)
+{
+    m_entryPoints.add(entryPoint);
+}
+
+
+// ComponentType
 
 ComponentType::ComponentType(Linkage* linkage)
     : m_linkage(linkage)
 {}
 
+ComponentType* asInternal(slang::IComponentType* inComponentType)
+{
+    // Note: we use a `queryInterface` here instead of just a `static_cast`
+    // to ensure that the `IComponentType` we get is the preferred/canonical
+    // one, which shares its address with the `ComponentType`.
+    //
+    // TODO: An alternative choice here would be to have a "magic" IID that
+    // we pass into `queryInterface` that returns the `ComponentType` directly
+    // (without even `addRef`-ing it).
+    //
+    ComPtr<slang::IComponentType> componentType;
+    inComponentType->queryInterface(IID_IComponentType, (void**) componentType.writeRef());
```

Comment:
Hmm. This assumes the inComponentType has an external ref count. It should probably check that. I mean it probably will... but there is potential for kaboom.

Better if we had the 'cast magic' where a UUID can return say RefObject* (and it does so without the refcount) you can do this without worrying about refcounts either. It would also allow the use of dynamic_cast (or some other mechanism) to sure the cast is to the impl.

Reply:
In this case it *must* have an external ref count because otherwise you are passing in an already-released pointer, no?

I agree that the magic cast would probably be better (hence the `TODO` comment). Would be a nice cleanup, but I'm ultimately just trying to get things working well enough to get back to my example program.